### PR TITLE
Add new validator options for running in more restrictive environments

### PR DESCRIPTION
**Pull request by mvines**
* `--gossip-pull-validator PUBKEY`: analogous to `--repair-validator` but for gossip pull requests.  That is, only pull from the explicitly provided validators
* `--restricted-repair-only-mode`: advertise no ports over gossip.  This will prevent the rest of the cluster from pushing packets to the validator, and essentially require that it continually repair only.   Useful in a setup where the validator will be mostly firewalled off from the rest of the cluster, to prevent the rest of the cluster from transmitting into a black hole

TODO:
* [x] Add some tests

### CODE REVIEW

**Comment by t-nelson** [comment]
Any plans to backport or should I prepare some stubs?

**Comment by mvines** [comment]
Definitely will backport all the way to 1.2.  I'm testing it out now.  Also need to add some exchange documentation to this PR before landing

**Comment by mvines** [comment]
@sakridge - mind taking a look at this for anything particularly offensive?  I'll add a bit of tests soon

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -1400,14 +1404,24 @@ impl ClusterInfo {
...
+        gossip_validators: Option<&HashSet<Pubkey>>,
+    ) -> Vec<(SocketAddr, Protocol)> {
         let self_id = self.id();
         let (_, push_messages) = self
             .time_gossip_write_lock("new_push_requests", &self.stats.new_push_requests)
             .new_push_messages(timestamp());
         let messages: Vec<_> = push_messages
             .into_iter()
             .filter_map(|(peer, messages)| {
+                if let Some(gossip_validators) = gossip_validators {
```
Comment (sakridge) [inline core/src/cluster_info.rs:1418]: I suspect limiting it in `crds_gossip_push::push_options` and `crds_gossip_pull::pull_options` might be the better/more efficient way. With this you are essentially considering the entire network and then throwing away packets that don't match, but you really could be forming your MST around the nodes that you can actually talk to and sending all your traffic to them.

Consider if you have 200 potential nodes, and only 2 in your trusted set. Your message generation would only be 1% efficient.

@behzadnouri @carllin if they have any opinions as well.

Comment (mvines) [reply core/src/cluster_info.rs:1418]: Yeah, makes sense.  I got shy about plumbing this into crds_gossip and crds_gossip_push.  I'll channel my inner 🦁 more.

Comment (carllin) [reply core/src/cluster_info.rs:1418]: limiting in ` crds_gossip_push::push_options` and `crds_gossip_pull::pull_options` makes sense to me!

Comment (mvines) [reply core/src/cluster_info.rs]: Done

**Comment by mvines** [comment]
Tests added

File: validator/src/main.rs [hunk 2]
```diff
@@ -1232,26 +1285,34 @@ pub fn main() {
...
+            udp_sockets.extend(node.sockets.tvu.iter());
+            udp_sockets.extend(node.sockets.broadcast.iter());
+            udp_sockets.extend(node.sockets.retransmit_sockets.iter());
+        }
+        if ContactInfo::is_valid_address(&node.info.tvu_forwards) {
+            udp_sockets.extend(node.sockets.tvu_forwards.iter());
+        }
 
         let mut tcp_listeners = vec![];
-        if !private_rpc {
```
Comment (ryoqun) [inline validator/src/main.rs:1248]: :eyes: https://github.com/solana-labs/solana/pull/12239

**Pull request merged**